### PR TITLE
feat(pkg/provenance): add OpenPGP signatures

### PR DESCRIPTION
**Pull request by technosophos**
This adds support for OpenPGP signatures containing provenance data.
Such information can be used to verify the integrity of a Chart by
testing that its file hash, metadata, and images are correct.

This first PR does not contain all of the tooling necessary for
end-to-end chart integrity. It contains just the library.

See #983

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/helm/988)

### CODE REVIEW

File: cmd/helm/list_test.go [hunk 1]
```diff
@@ -56,9 +56,7 @@ func TestListCmd(t *testing.T) {
 			rels: tt.resp,
 		}
 		cmd := newListCmd(c, &buf)
-		for flag, value := range tt.flags {
-			cmd.Flags().Set(flag, value)
-		}
+		setFlags(cmd, tt.flags)
```
Comment (adamreese) [inline cmd/helm/list_test.go]: flags can be added as args, I deleted the flags arg a while ago, it is a zombie that cannot be killed

File: cmd/helm/package.go [hunk 2]
```diff
@@ -37,30 +43,56 @@ Chart.yaml file, and (if found) build the current directory into a chart.
 Versioned chart archives are used by Helm package repositories.
 `
 
-var save bool
+const (
+	envSigningKey = "HELM_SIGNING_KEY"
+	envKeyring    = "HELM_KEYRING"
```
Comment (adamreese) [inline cmd/helm/package.go]: will these be used somewhere?

File: cmd/helm/fetch.go [hunk 3]
```diff
@@ -27,65 +30,176 @@ import (
...
+There are options for unpacking the chart after download. This will create a
+directory for the chart and uncomparess into that directory.
+
+If the --verify flag is specified, the requested chart MUST have a provenance
+file, and MUST pass the verification process. Failure in any part of this will
+result in an error, and the chart will not be saved locally.
+`
+
 var untarFile bool
 var untarDir string
```
Comment (adamreese) [inline cmd/helm/fetch.go]: delete us

File: cmd/helm/verify_test.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			err:    false,
+		},
+	}
+
+	for _, tt := range tests {
+		b := bytes.NewBuffer(nil)
+		vc := newVerifyCmd(b)
+		vc.ParseFlags(tt.flags)
+		err := vc.RunE(vc, tt.args)
+		if tt.err == true {
```
Comment (adamreese) [inline cmd/helm/verify_test.go]: `if tt.err {...}`

File: pkg/provenance/sign.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+	"k8s.io/helm/pkg/chartutil"
+	hapi "k8s.io/helm/pkg/proto/hapi/chart"
+)
+
+var defaultPGPConfig = packet.Config{
+	DefaultHash: crypto.SHA512,
+}
+
+// SumCollection represents a collecton of file and image checksums.
```
Comment (adamreese) [inline pkg/provenance/sign.go]: s/collecton/collection/

File: cmd/helm/list_test.go (hunk 1 above)
Comment (technosophos) [inline cmd/helm/list_test.go]: I don't understand what you are suggesting I do here. Simply putting `args: []string{"--long"}` doesn't work.

**Pull request merged**